src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -140,18 +147,49 @@ private static void checkpointRestore1() throws
             }
         }
 
+        RestoreException restoreException = null;
         try {
             globalContext.afterRestore(null);
         } catch (RestoreException re) {
             if (checkpointException == null) {
-                throw re;
+                restoreException = re;
+            } else {
+                for (Throwable t : re.getSuppressed()) {
+                    checkpointException.addSuppressed(t);
+                }
             }
-            for (Throwable t : re.getSuppressed()) {
-                checkpointException.addSuppressed(t);
+        }
+
+        if (newArguments != null && newArguments.length() > 0) {
+            String[] args = newArguments.split(" ");
+            if (args.length > 0) {
+                try {
+                    Class<?> newMainClass = Class.forName(args[0], false,
+                        ClassLoader.getSystemClassLoader());
+                    Method newMain = newMainClass.getDeclaredMethod("main",
+                        String[].class);
+                    newMain.setAccessible(true);
+                    newMain.invoke(null,
+                        (Object)Arrays.copyOfRange(args, 1, args.length));
+                } catch (ClassNotFoundException    |
+                         InvocationTargetException |
+                         NoSuchMethodException     |
+                         IllegalAccessException e) {
```

Comment:
Thanks for adapting this to be more specific on what it catches.

I don't think it's quite right yet though as the current code will allow exceptions thrown by the `newMain` method (and its callees) to propagate past the checkpoint spot and makes the caller of the checkpoint code need to handle them.  

At the time `Core.checkpointRestore();` is called, we may have stack that looks llike:
```
TOS
Core.checkpointRestore();
Foo.bar();
Foo.foobar();
SomeOtherClass.method();
OriginalClass.main();
```

And when we restore, we load an execute a new `main()` method as though it was called where `Core.checkpointRestore();` was previously on the stack resulting in:
```
TOS
newMain.main();
Core.checkpointRestore();
Foo.bar();
Foo.foobar();
SomeOtherClass.method();
OriginalClass.main();
```

So exceptions thrown by code called from `newMain` should not propagate past `Core.checkpointRestore();` without being wrapped in a `RestoreException`.  `Error`-subclasses should propagate.

I think the code should be refactored to something like:
```
               } catch(Exception t) {
                   assert checkpointException == null :
                        "should not have new arguments";
                    if (restoreException == null) {
                        restoreException = new RestoreException();
                    }
                    restoreException.addSuppressed(e);
                }
```
as that correctly catches all Exceptions but lets the Errors propagate past

Reply:
> the current code will allow exceptions thrown by the `newMain` method (and its callees) to propagate past the checkpoint spot and makes the caller of the checkpoint code need to handle them.

An exception from the newMain will be wrapped in InvocationTargetException, which is caught and is suppressed by the RestoreException. Right? Otherwise, this java code won't compile, as checkpointRestore is declared to throw only {Checkpoint,Restore}Exception.

Reply:
I had to double check with `jshell` as I thought runtime exceptions would propagate past.  Turns out your 100% right about the all Exceptions/Errors/Throwables being wrapped with `InvocationTargetException`.